### PR TITLE
Refine time index of Flows

### PR DESCRIPTION
**Pull request by p-snft**
As different types of periods will be used for multi-period optimisation (long-term investment planing) and working with aggregated time-series (e.g. using tsam), I suggest to change the indexing. This PR implements Flows that use timesteps [int] instead of timeindex currently ([int, int]).

### CODE REVIEW

**Comment by p-snft** [comment]
Maybe, we should split this into several PRs and merge this as a first step.

File: src/oemof/solph/components/_offset_converter.py [hunk 1]
```diff
@@ -365,16 +366,14 @@ def __init__(
         inputs,
         outputs,
         label=None,
-        conversion_factors=None,
-        normed_offsets=None,
+        coefficients=None,
```
Comment (fwitte) [inline src/oemof/solph/components/_offset_converter.py:369]: Is the deletion and addition here intended? Or some kind of relict from a merge?

File: src/oemof/solph/components/_offset_converter.py [hunk 2]
```diff
@@ -365,16 +366,14 @@ def __init__(
...
         custom_attributes=None,
     ):
         super().__init__(
             label=label,
             inputs=inputs,
             outputs=outputs,
-            conversion_factors=conversion_factors,
-            normed_offsets=normed_offsets,
+            coefficients=coefficients,
```
Comment (fwitte) [inline src/oemof/solph/components/_offset_converter.py:376]: Deletion and insertion intended here?

File: src/oemof/solph/flows/_non_convex_flow_block.py [hunk 3]
```diff
@@ -743,17 +743,11 @@ def _positive_gradient_flow_constraint(_):
                             m.flow[
                                 i,
                                 o,
-                                m.TIMEINDEX[index][0],
-                                m.TIMEINDEX[index][1],
+                                m.TIMESTEPS[index],
```
Comment (fwitte) [inline src/oemof/solph/flows/_non_convex_flow_block.py:746]: Kind of not directly related, but why does line `740` iterate over a numeric `index` instead of `t` as is done everywhere else? I guess there is a good reason for it...

File: src/oemof/solph/flows/_non_convex_flow_block.py [hunk 4]
```diff
@@ -790,33 +782,28 @@ def _positive_gradient_flow_constraint(_):
         def _negative_gradient_flow_constraint(_):
             r"""Rule definition for negative gradient constraint."""
             for i, o in self.NEGATIVE_GRADIENT_FLOWS:
-                for index in range(1, len(m.TIMEINDEX) + 1):
-                    if m.TIMEINDEX[index][1] > 0:
+                for index in range(1, len(m.TIMESTEPS) + 1):
+                    if m.TIMESTEPS[index] > 0:
```
Comment (fwitte) [inline src/oemof/solph/flows/_non_convex_flow_block.py:786]: Same here, why is there no iteration on the timesteps but over the index?

**Comment by fwitte** [comment]
Thank you for the work @p-snft, since the tests are working things look fine for me. However, I am not very knowledgeable yet on how multi-period would work and cannot really check the changes for content. Also, I did not check the lp-files changes... That is simply too much, so #1079 is really welcome! :)

File: src/oemof/solph/components/_offset_converter.py (hunk 1 above)
Comment (p-snft) [reply src/oemof/solph/components/_offset_converter.py:369]: Eventually, I made a mistake when merging. That would mean that i also made another one when reviewing #1068. (The `OffsetTransformer` as a transitional wrapper should offer the old API, shouldn't it?)

Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py:369]: It does: https://github.com/oemof/oemof-solph/blob/80298509a7e8e485cd2b2c35cf2fcfa0fcfbf399/src/oemof/solph/components/_offset_converter.py#L117-L119, there also is a test: https://github.com/oemof/oemof-solph/blob/80298509a7e8e485cd2b2c35cf2fcfa0fcfbf399/tests/test_components/test_offset_converter.py#L420

Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py:369]: And, in this branch it is also available: https://github.com/oemof/oemof-solph/blob/revision/refine_timeindex/tests/test_components/test_offset_converter.py

Therefore, I guess that those keywords do not have any effect (!?)...

File: src/oemof/solph/flows/_non_convex_flow_block.py (hunk 3 above)
Comment (p-snft) [reply src/oemof/solph/flows/_non_convex_flow_block.py:746]: To my understanding, this is because of the `positive_gradient_constr` needing the `TIMEINDEX` (including periods) to be allowed to consider investment (periods) in the ramping. (@jokochems might now better.) There is probably a more elegant way to do this, but I did not want to touch this now.

File: src/oemof/solph/flows/_non_convex_flow_block.py (hunk 4 above)
Comment (p-snft) [reply src/oemof/solph/flows/_non_convex_flow_block.py:786]: Here, I do not see the point I made before. Maybe, the current implementation is faulty?

File: src/oemof/solph/components/_offset_converter.py (hunk 1 above)
Comment (p-snft) [reply src/oemof/solph/components/_offset_converter.py:369]: As the keywords `conversion_factors` and `normed_offsets` are not part of the old API, I thought they were in the merge by mistake (or because I changed something). In my opinion, it does not make sense to have an `OffsetTransformer` with these keywords.

Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py:369]: Ah that is on the transformer... I overlooked that, then all is fine! :)

File: src/oemof/solph/flows/_non_convex_flow_block.py (hunk 4 above)
Comment (fwitte) [reply src/oemof/solph/flows/_non_convex_flow_block.py:786]: I mean it works, it was just a side comment. I would say: merge the feature and maybe also make a release to ship the latest changes to the users.

File: src/oemof/solph/flows/_non_convex_flow_block.py (hunk 3 above)
Comment (fwitte) [reply src/oemof/solph/flows/_non_convex_flow_block.py:746]: Maybe also something to be discussed in a separate issue to not delay the progress too much

**Review by fwitte — APPROVED** [review]
All remaining issues are side issues, which can be handled in a different PR

**Review by jokochems — APPROVED** [review]
Thank you, @p-snft!

It has always bugged me that the multi-period feature brought along this inconsistency between indexing flows and other variables. 

I thought about indexing every not investment-related variable with the `TIMEINDEX` set, but getting rid of the `TIMEINDEX` and indexing flows in `TIMESTEPS` instead, seems more user-friendly and intuitive to me. I just did not come up with that because I seem to have been too routine-blinded.

Preserving periods where needed can be achieved by using the `TIMEINDEX` set for iterations, as you did. Well actually you just did not touch upon that and kept it as is.

Long story short: I am happy to approve this as well. I only had one minor suggestion for renaming.

P.S.: It has been a while since I looked into the code base, but I still have the feeling of understanding it, which is a very good thing. ;-)

File: src/oemof/solph/processing.py [hunk 5]
```diff
@@ -277,8 +277,8 @@ def results(model, remove_last_time_point=False):
         )
         for bus, timeindex in grouped:
             duals = [
-                model.dual[model.BusBlock.balance[bus, p, t]]
-                for _, p, t in timeindex
+                model.dual[model.BusBlock.balance[bus, t]]
+                for _, t in timeindex
```
[the remark this answers is not in the document]
Comment (p-snft) [reply src/oemof/solph/processing.py:281]: It should be a timeindex, as it contains something (we don't care about) an the time step `_, t`. Or am I wrong?

File: src/oemof/solph/components/_generic_storage.py [hunk 6]
```diff
@@ -1258,7 +1258,7 @@ def _storage_investvar_bound_rule(block, n, p):
         i = {n: [i for i in n.inputs][0] for n in group}
         o = {n: [o for o in n.outputs][0] for n in group}
 
-        reduced_periods_timesteps = [(p, t) for (p, t) in m.TIMEINDEX if t > 0]
+        reduced_periods_timeindex = [(p, t) for (p, t) in m.TIMEINDEX if t > 0]
```
[the remark this answers is not in the document]
Comment (p-snft) [reply src/oemof/solph/components/_generic_storage.py:1261]: To my understanding, this removes all indexes with t <= 0. If we can assume the index to be sorted ans starting with 0, this can be more efficient, sure.

File: src/oemof/solph/flows/_non_convex_flow_block.py (hunk 3 above)
Comment (p-snft) [reply src/oemof/solph/flows/_non_convex_flow_block.py:746]: If it is really the periods, I did something wrong, as `m.TIMESTEPS[p]` does not make sense.

File: src/oemof/solph/components/_generic_storage.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jokochems) [reply src/oemof/solph/components/_generic_storage.py:1261]: Okay, I think, this can be resolved as it was rather a remark, not a demand for a change.

**Comment by jokochems** [comment]
@p-snft I'm fine with you addressing the two comments according to your preferences (as I already approved it) and then merging it rightaways. Thank you very much!

File: src/oemof/solph/processing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (p-snft) [reply src/oemof/solph/processing.py:281]: With some days of distance: You are right.

**Comment by p-snft** [comment]
As the gradient constraint has problems anyway, I will not touch it in this PR.
